File: scripts/parity/schema/node/canonicalize.test.ts
import { describe, expect, it } from "vitest";
import { canonicalize } from "./canonicalize.js";
import type { NativeDump } from "./canonicalize.js";

describe("canonicalize", () => {
  it("maps a trivial fixture to canonical form", () => {
    const native: NativeDump = {
      users: {
        columns: [
          {
            name: "id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "email",
            sqlType: "TEXT",
            primaryKey: false,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "name",
            sqlType: "TEXT",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "score",
            sqlType: "REAL",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "avatar",
            sqlType: "BLOB",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "created_at",
            sqlType: "DATETIME",
            primaryKey: false,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "active",
            sqlType: "INTEGER",
            primaryKey: false,
            null: false,
            default: "1",
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: ["id"],
      },
    };

    const result = canonicalize(native);

    expect(result.version).toBe(1);
    expect(result.tables).toHaveLength(1);
    const [table] = result.tables;
    expect(table!.name).toBe("users");
    expect(table!.primaryKey).toBe("id");
    expect(table!.columns.map((c) => c.name)).toEqual([
      "id",
      "email",
      "name",
      "score",
      "avatar",
      "created_at",
      "active",
    ]);
    expect(table!.columns.find((c) => c.name === "id")!.type).toBe("integer");
    expect(table!.columns.find((c) => c.name === "score")!.type).toBe("float");
    expect(table!.columns.find((c) => c.name === "avatar")!.type).toBe("binary");
    expect(table!.columns.find((c) => c.name === "created_at")!.type).toBe("datetime");
    expect(table!.columns.find((c) => c.name === "active")!.default).toBe(1);
    expect(table!.indexes).toHaveLength(0);
  });

  it("preserves column declaration order", () => {
    const native: NativeDump = {
      things: {
        columns: [
          {
            name: "z",
            sqlType: "TEXT",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "a",
            sqlType: "TEXT",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "m",
            sqlType: "TEXT",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.columns.map((c) => c.name)).toEqual(["z", "a", "m"]);
  });

  it("sorts tables by name", () => {
    const native: NativeDump = {
      zebra: {
        columns: [
          {
            name: "id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: ["id"],
      },
      apple: {
        columns: [
          {
            name: "id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: ["id"],
      },
    };
    const result = canonicalize(native);
    expect(result.tables.map((t) => t.name)).toEqual(["apple", "zebra"]);
  });

  it("filters schema_migrations and ar_internal_metadata", () => {
    const native: NativeDump = {
      users: {
        columns: [
          {
            name: "id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: ["id"],
      },
      schema_migrations: {
        columns: [
          {
            name: "version",
            sqlType: "TEXT",
            primaryKey: false,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
      ar_internal_metadata: {
        columns: [
          {
            name: "key",
            sqlType: "TEXT",
            primaryKey: false,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
    };
    const result = canonicalize(native);
    expect(result.tables.map((t) => t.name)).toEqual(["users"]);
  });

  it("filters sqlite_autoindex_* and sorts remaining indexes by name", () => {
    const native: NativeDump = {
      posts: {
        columns: [
          {
            name: "id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [
          { name: "sqlite_autoindex_posts_1", columns: ["title"], unique: true },
          { name: "idx_posts_z", columns: ["created_at"], unique: false },
          { name: "idx_posts_a", columns: ["author_id"], unique: false },
        ],
        primaryKeyColumns: ["id"],
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.indexes.map((i) => i.name)).toEqual(["idx_posts_a", "idx_posts_z"]);
  });

  it("represents composite PK as tuple", () => {
    const native: NativeDump = {
      taggings: {
        columns: [
          {
            name: "tag_id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "taggable_id",
            sqlType: "INTEGER",
            primaryKey: true,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: ["tag_id", "taggable_id"],
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.primaryKey).toEqual(["tag_id", "taggable_id"]);
  });

  it("uses primaryKeyColumns order, not column declaration order, for composite PK", () => {
    // Simulates PRIMARY KEY (b, a) where a appears first in the table definition.
    // Rails preserves the PK constraint order via PRAGMA table_info `pk` position.
    const native: NativeDump = {
      items: {
        columns: [
          {
            name: "a",
            sqlType: "INTEGER",
            primaryKey: true,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "b",
            sqlType: "INTEGER",
            primaryKey: true,
            null: false,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: ["b", "a"], // PK constraint order: b first, then a
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.primaryKey).toEqual(["b", "a"]);
  });

  it("represents no-PK table", () => {
    const native: NativeDump = {
      logs: {
        columns: [
          {
            name: "message",
            sqlType: "TEXT",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.primaryKey).toBeNull();
  });

  it("coerces numeric string defaults", () => {
    const native: NativeDump = {
      t: {
        columns: [
          {
            name: "count",
            sqlType: "INTEGER",
            primaryKey: false,
            null: false,
            default: "0",
            limit: null,
            precision: null,
            scale: null,
          },
          {
            name: "rate",
            sqlType: "REAL",
            primaryKey: false,
            null: false,
            default: "1.5",
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.columns[0]!.default).toBe(0);
    expect(table!.columns[1]!.default).toBe(1.5);
  });

  it("coerces quoted string defaults", () => {
    const native: NativeDump = {
      t: {
        columns: [
          {
            name: "status",
            sqlType: "TEXT",
            primaryKey: false,
            null: false,
            default: "'active'",
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
    };
    const [table] = canonicalize(native).tables;
    expect(table!.columns[0]!.default).toBe("active");
  });

  it("throws on unknown SQL type", () => {
    const native: NativeDump = {
      t: {
        columns: [
          {
            name: "x",
            sqlType: "UNKNOWNTYPE",
            primaryKey: false,
            null: true,
            default: null,
            limit: null,
            precision: null,
            scale: null,
          },
        ],
        indexes: [],
        primaryKeyColumns: [],
      },
    };
    expect(() => canonicalize(native)).toThrow(/unknown SQL type "UNKNOWNTYPE"/);
  });
});

File: scripts/parity/schema/ruby/canonicalize_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "canonicalize"

class CanonicalizeTest < Minitest::Test
  # Build a minimal NativeTable hash for test fixtures
  def col(name, ar_type, null: true, default: nil, limit: nil, precision: nil, scale: nil)
    { name: name, ar_type: ar_type, null: null, default: default,
      limit: limit, precision: precision, scale: scale }
  end

  def table(columns:, indexes: [], primary_key_columns: [])
    { columns: columns, indexes: indexes, primary_key_columns: primary_key_columns }
  end

  # --- type mapping ---

  def test_maps_trivial_fixture_columns
    native = {
      "users" => table(
        columns: [
          col("id",         :integer,  null: true),
          col("email",      :text,     null: false),
          col("name",       :text,     null: true),
          col("score",      :float,    null: true),
          col("avatar",     :binary,   null: true),
          col("created_at", :datetime, null: false),
          col("active",     :integer,  null: false, default: 1),
        ],
        primary_key_columns: ["id"],
      ),
    }

    result = Canonicalize.call(native)

    assert_equal 1, result["version"]
    assert_equal 1, result["tables"].length
    t = result["tables"][0]
    assert_equal "users", t["name"]
    assert_equal "id", t["primaryKey"]
    assert_equal %w[id email name score avatar created_at active], t["columns"].map { |c| c["name"] }
    assert_equal "integer",  t["columns"][0]["type"]
    assert_equal "float",    t["columns"][3]["type"]
    assert_equal "binary",   t["columns"][4]["type"]
    assert_equal "datetime", t["columns"][5]["type"]
    assert_equal 1,          t["columns"][6]["default"]
    assert_equal [],         t["indexes"]
  end

  # --- column ordering ---

  def test_preserves_column_declaration_order
    native = {
      "things" => table(columns: [col("z", :text), col("a", :text), col("m", :text)]),
    }
    t = Canonicalize.call(native)["tables"][0]
    assert_equal %w[z a m], t["columns"].map { |c| c["name"] }
  end

  # --- table sorting ---

  def test_sorts_tables_by_name
    native = {
      "zebra" => table(columns: [col("id", :integer)], primary_key_columns: ["id"]),
      "apple" => table(columns: [col("id", :integer)], primary_key_columns: ["id"]),
    }
    names = Canonicalize.call(native)["tables"].map { |t| t["name"] }
    assert_equal %w[apple zebra], names
  end

  # --- filters ---

  def test_filters_schema_migrations_and_ar_internal_metadata
    native = {
      "users"                => table(columns: [col("id", :integer)], primary_key_columns: ["id"]),
      "schema_migrations"    => table(columns: [col("version", :string, null: false)]),
      "ar_internal_metadata" => table(columns: [col("key", :string, null: false)]),
    }
    names = Canonicalize.call(native)["tables"].map { |t| t["name"] }
    assert_equal ["users"], names
  end

  def test_filters_sqlite_autoindex_and_sorts_remaining_indexes
    native = {
      "posts" => table(
        columns: [col("id", :integer)],
        primary_key_columns: ["id"],
        indexes: [
          { name: "sqlite_autoindex_posts_1", columns: ["title"],      unique: true,  where: nil },
          { name: "idx_posts_z",              columns: ["created_at"], unique: false, where: nil },
          { name: "idx_posts_a",              columns: ["author_id"],  unique: false, where: nil },
        ],
      ),
    }
    t = Canonicalize.call(native)["tables"][0]
    assert_equal %w[idx_posts_a idx_posts_z], t["indexes"].map { |i| i["name"] }
  end

  # --- primary key shapes ---

  def test_single_column_pk
    native = { "t" => table(columns: [col("id", :integer)], primary_key_columns: ["id"]) }
    assert_equal "id", Canonicalize.call(native)["tables"][0]["primaryKey"]
  end

  def test_composite_pk_as_array_in_position_order
    native = {
      "taggings" => table(
        columns: [col("tag_id", :integer, null: false), col("taggable_id", :integer, null: false)],
        primary_key_columns: ["tag_id", "taggable_id"],
      ),
    }
    assert_equal ["tag_id", "taggable_id"], Canonicalize.call(native)["tables"][0]["primaryKey"]
  end

  def test_pk_position_order_overrides_column_declaration_order
    # Simulates PRIMARY KEY (b, a) where a appears first in the table definition.
    native = {
      "items" => table(
        columns: [col("a", :integer, null: false), col("b", :integer, null: false)],
        primary_key_columns: ["b", "a"],
      ),
    }
    assert_equal ["b", "a"], Canonicalize.call(native)["tables"][0]["primaryKey"]
  end

  def test_no_pk_table
    native = { "logs" => table(columns: [col("message", :text)]) }
    assert_nil Canonicalize.call(native)["tables"][0]["primaryKey"]
  end

  # --- default coercion ---

  def test_integer_default
    native = { "t" => table(columns: [col("count", :integer, null: false, default: 0)]) }
    assert_equal 0, Canonicalize.call(native)["tables"][0]["columns"][0]["default"]
  end

  def test_string_default
    native = { "t" => table(columns: [col("status", :string, null: false, default: "active")]) }
    assert_equal "active", Canonicalize.call(native)["tables"][0]["columns"][0]["default"]
  end

  def test_nil_default
    native = { "t" => table(columns: [col("note", :text, null: true, default: nil)]) }
    assert_nil Canonicalize.call(native)["tables"][0]["columns"][0]["default"]
  end

  # --- error cases ---

  def test_raises_on_unknown_ar_type
    native = { "t" => table(columns: [col("x", :unknowntype)]) }
    err = assert_raises(RuntimeError) { Canonicalize.call(native) }
    assert_match(/unknown AR type.*unknowntype/, err.message)
  end

  def test_raises_on_index_with_no_columns
    native = {
      "t" => table(
        columns: [col("id", :integer)],
        indexes: [{ name: "bad_idx", columns: [], unique: false, where: nil }],
      ),
    }
    err = assert_raises(RuntimeError) { Canonicalize.call(native) }
    assert_match(/no columns/, err.message)
  end
end
